graph-http/src/async_client/client.rs

```diff
@@ -155,326 +159,214 @@ impl From<Url> for AsyncClient {
-    where
-        F: Fn(&GraphUrl) + Sync,
-    {
-        f(&self.client.lock().await.url)
-    }
-
-    async fn inner_url_mut<F>(&self, f: F)
-    where
-        F: Fn(&mut GraphUrl) + Sync,
-    {
-        f(&mut self.client.lock().await.url)
-    }
-
-    async fn inner_debug(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
-        self.client.lock().await.fmt(f)
-    }
-
-    async fn inner_registry<F>(&self, f: F)
-    where
-        F: Fn(&mut Handlebars) + Sync,
-    {
-        f(&mut self.client.lock().await.registry)
-    }
-
-    async fn inner_render_template(&self, template: &str, json: &serde_json::Value) -> String {
-        self.client
-            .lock()
-            .await
-            .registry
-            .render_template(template, json)
-            .unwrap()
-    }
-
-    async fn inner_register_ident_helper(&self, resource_identity: ResourceIdentity) {
-        self.client
-            .lock()
-            .await
-            .register_ident_helper(resource_identity);
-    }
-
-    async fn inner_extend_path(&self, path: &[&str]) {
-        self.client.lock().await.url.extend_path(path);
-    }
-
-    async fn inner_set_request(
-        &self,
-        req_att: Vec<RequestAttribute<reqwest::Body, reqwest::multipart::Form>>,
-    ) -> GraphResult<()> {
-        let mut client = self.client.lock().await;
-        for att in req_att {
-            match att {
-                RequestAttribute::Token(token) => client.token = token,
-                RequestAttribute::Ident(ident) => client.ident = ident,
-                RequestAttribute::Url(url) => client.url = url,
-                RequestAttribute::Method(method) => client.method = method,
-                RequestAttribute::Body(body) => client.body = Some(body),
-                RequestAttribute::BodyFile(path) => {
-                    let buffer = tokio::fs::read_to_string(path).await?;
-                    client.body = Some(buffer.into());
-                }
-                RequestAttribute::Headers(headers) => client.headers = headers,
-                RequestAttribute::ClearHeaders => client.headers.clear(),
-                RequestAttribute::Download(path) => client.download_dir = Some(path),
-                RequestAttribute::Upload(path) => client.upload_session_file = Some(path),
-                RequestAttribute::Form(form) => client.form = Some(form),
-                RequestAttribute::RequestType(req_type) => client.req_type = req_type,
-            }
-        }
-        Ok(())
+    pub fn clone_inner(&self) -> Arc<Mutex<AsyncClient>> {
+        Arc::clone(&self.client)
     }
 
     pub async fn download(&self) -> AsyncDownload {
-        self.client.lock().await.download()
+        self.client.lock().unwrap().download()
     }
 
     pub async fn upload_session(&self) -> GraphResult<UploadSessionClient<AsyncHttpClient>> {
-        self.client.lock().await.upload_session().await
+        self.client.lock().unwrap().upload_session().await
```

Comment:
I thought this line might cause a deadlock but I tested it and everything is fine, even with high concurrency.

Reply:
Thats good. I am still not a fan of all these unwraps, but I cant remember if its best practice to unwrap on mutexes or to return them in a result. I assume it would be fine if it wasnt an issue with threads but I think thats why your suppose to unwrap those. I need to look into this again cause its been a while.

Reply:
I switched it to parking_lot::Mutex which Tokio also recommends. If we run into this one Ill switch back but this should work fine.

Reply:
@DevLazio You work more with requests that highlight the issues with concurrency more so than I do so if you have some good ways of testing it then please do create/update the tests on this. Its probably needed just to help prevent bugs like this in the future.

Reply:
Sorry for the late response.
Everything works great on my end with your latests changes.

For the tests, the only thing I can think of is increasing the number of concurrent requests in [this test](https://github.com/sreeise/graph-rs/commit/b8d8240809a299714b5f2e8e4dbc0f762cb922bb), in the hope it will be enough to trigger a problem.